File: src/core/feedback.py
"""Feedback form seam: validate POST → journal (local) / mail (host).

Mirrors ``wp-content/plugins/ichnm-site/includes/feedback.php`` (ticket 30).
No CRM; recipient is constant / option / default institute address.
"""

from __future__ import annotations

import re
from typing import Any, Literal, Mapping

DEFAULT_FEEDBACK_TO = "ichnm@ichnm.by"
MAIL_SUBJECT = "Обратная связь с сайта ИХНМ"

_EMAIL_RE = re.compile(
    r"^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+@"
    r"[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?"
    r"(?:\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)+$"
)

DeliveryMode = Literal["journal", "mail"]


def _trim(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def _is_email(value: str) -> bool:
    if not value or len(value) > 180:
        return False
    return bool(_EMAIL_RE.match(value))


def validate_feedback(raw: Mapping[str, Any]) -> dict[str, Any]:
    """Return ``{ok, name, email, message}`` or ``{ok: False, error: 'fields'}``."""
    name = _trim(raw.get("name"))
    email = _trim(raw.get("email")).lower()
    message = _trim(raw.get("message"))
    if not name or len(name) > 120:
        return {"ok": False, "error": "fields"}
    if not message or len(message) > 5000:
        return {"ok": False, "error": "fields"}
    if not _is_email(email):
        return {"ok": False, "error": "fields"}
    return {"ok": True, "name": name, "email": email, "message": message}


def resolve_delivery_mode(
    *,
    override: str | None = None,
    environment: str | None = None,
    home_url: str | None = None,
) -> DeliveryMode:
    """Choose journal (Docker/local) vs mail (PHP host).

    Localhost home URL wins over WordPress's default ``production``
    environment type (common on Docker images).
    """
    if override is not None:
        mode = override.strip().lower()
        if mode not in ("journal", "mail"):
            raise ValueError(f"unknown feedback delivery mode: {override!r}")
        return mode  # type: ignore[return-value]

    url = (home_url or "").strip().lower()
    if "localhost" in url or "127.0.0.1" in url:
        return "journal"

    env = (environment or "").strip().lower()
    if env in ("local", "development"):
        return "journal"
    if env in ("production", "staging"):
        return "mail"

    return "mail"


def resolve_recipient(
    *,
    constant: str | None = None,
    option: str | None = None,
    default: str = DEFAULT_FEEDBACK_TO,
) -> str:
    """Prefer ``ICHNM_FEEDBACK_TO``, then option ``ichnm_feedback_to``, then default."""
    for candidate in (constant, option, default):
        value = _trim(candidate)
        if value and _is_email(value):
            return value.lower()
    return DEFAULT_FEEDBACK_TO


def feedback_mail_body(name: str, email: str, message: str) -> str:
    return f"Имя: {name}\nEmail: {email}\n\n{message}\n"


def plan_feedback_delivery(
    validated: Mapping[str, Any],
    *,
    mode: DeliveryMode,
    recipient: str,
) -> dict[str, Any]:
    """Pure plan: validation result + mode → journal or mail action (no I/O)."""
    if not validated.get("ok"):
        return {"ok": False, "error": validated.get("error") or "fields"}

    body = feedback_mail_body(
        str(validated["name"]),
        str(validated["email"]),
        str(validated["message"]),
    )
    to = resolve_recipient(constant=recipient)
    return {
        "ok": True,
        "mode": mode,
        "action": mode,
        "recipient": to,
        "subject": MAIL_SUBJECT,
        "body": body,
        "name": validated["name"],
        "email": validated["email"],
        "message": validated["message"],
    }

File: wp-content/plugins/ichnm-site/includes/feedback.php
<?php
/**
 * Feedback delivery seam (ticket 30).
 *
 * Validate POST → journal on local Docker contour; wp_mail on PHP host.
 * Recipient: constant ICHNM_FEEDBACK_TO, else option ichnm_feedback_to, else default.
 * Pure helpers mirror src/core/feedback.py (pytest without Docker).
 */

if (!defined('ABSPATH')) {
    exit;
}

const ICHNM_FEEDBACK_DEFAULT_TO = 'ichnm@ichnm.by';
const ICHNM_FEEDBACK_MAIL_SUBJECT = 'Обратная связь с сайта ИХНМ';

/**
 * Soft email check aligned with the Python seam (does not require WP is_email).
 */
function ichnm_feedback_looks_like_email(string $email): bool
{
    if ($email === '' || strlen($email) > 180) {
        return false;
    }
    return (bool) preg_match(
        "/^[A-Za-z0-9.!#$%&'*+\/=?^_`{|}~-]+@"
        . "[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?"
        . "(?:\\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)+$/",
        $email
    );
}

/**
 * @param array{name?:string,email?:string,message?:string} $raw
 * @return array{ok:true,name:string,email:string,message:string}|array{ok:false,error:string}
 */
function ichnm_feedback_validate(array $raw): array
{
    $name = isset($raw['name']) ? trim((string) $raw['name']) : '';
    $email = isset($raw['email']) ? strtolower(trim((string) $raw['email'])) : '';
    $message = isset($raw['message']) ? trim((string) $raw['message']) : '';

    if ($name === '' || strlen($name) > 120) {
        return ['ok' => false, 'error' => 'fields'];
    }
    if ($message === '' || strlen($message) > 5000) {
        return ['ok' => false, 'error' => 'fields'];
    }
    if (!ichnm_feedback_looks_like_email($email)) {
        return ['ok' => false, 'error' => 'fields'];
    }
    if (function_exists('is_email') && !is_email($email)) {
        return ['ok' => false, 'error' => 'fields'];
    }

    return [
        'ok' => true,
        'name' => $name,
        'email' => $email,
        'message' => $message,
    ];
}

/**
 * @param array{override?:string,environment?:string,home_url?:string} $ctx
 */
function ichnm_feedback_delivery_mode(array $ctx = []): string
{
    if (defined('ICHNM_FEEDBACK_DELIVERY')) {
        $override = strtolower(trim((string) ICHNM_FEEDBACK_DELIVERY));
        if ($override === 'journal' || $override === 'mail') {
            return $override;
        }
    }
    if (!empty($ctx['override'])) {
        $override = strtolower(trim((string) $ctx['override']));
        if ($override === 'journal' || $override === 'mail') {
            return $override;
        }
    }

    // Localhost home wins over WP's default "production" (typical on Docker).
    $home = isset($ctx['home_url'])
        ? strtolower((string) $ctx['home_url'])
        : (function_exists('home_url') ? strtolower((string) home_url('/')) : '');
    if ($home !== '' && (str_contains($home, 'localhost') || str_contains($home, '127.0.0.1'))) {
        return 'journal';
    }

    $env = isset($ctx['environment'])
        ? strtolower(trim((string) $ctx['environment']))
        : (function_exists('wp_get_environment_type') ? strtolower((string) wp_get_environment_type()) : '');
    if (in_array($env, ['local', 'development'], true)) {
        return 'journal';
    }
    if (in_array($env, ['production', 'staging'], true)) {
        return 'mail';
    }

    return 'mail';
}

/**
 * @param array{constant?:string,option?:string} $ctx
 */
function ichnm_feedback_recipient(array $ctx = []): string
{
    $candidates = [];
    if (defined('ICHNM_FEEDBACK_TO')) {
        $candidates[] = (string) ICHNM_FEEDBACK_TO;
    }
    if (array_key_exists('constant', $ctx)) {
        $candidates[] = (string) $ctx['constant'];
    }
    if (array_key_exists('option', $ctx)) {
        $candidates[] = (string) $ctx['option'];
    } elseif (function_exists('get_option')) {
        $candidates[] = (string) get_option('ichnm_feedback_to', '');
    }
    $candidates[] = ICHNM_FEEDBACK_DEFAULT_TO;

    foreach ($candidates as $candidate) {
        $email = strtolower(trim($candidate));
        if ($email === '') {
            continue;
        }
        if (ichnm_feedback_looks_like_email($email)) {
            if (function_exists('is_email') && !is_email($email)) {
                continue;
            }
            return $email;
        }
    }

    return ICHNM_FEEDBACK_DEFAULT_TO;
}

function ichnm_feedback_mail_body(string $name, string $email, string $message): string
{
    return "Имя: {$name}\nEmail: {$email}\n\n{$message}\n";
}

/**
 * Pure plan after validation (no I/O).
 *
 * @param array{ok:bool,name?:string,email?:string,message?:string,error?:string} $validated
 * @return array{ok:true,mode:string,action:string,recipient:string,subject:string,body:string,name:string,email:string,message:string}|array{ok:false,error:string}
 */
function ichnm_feedback_plan(array $validated, string $mode, string $recipient): array
{
    if (empty($validated['ok'])) {
        return ['ok' => false, 'error' => (string) ($validated['error'] ?? 'fields')];
    }
    $name = (string) $validated['name'];
    $email = (string) $validated['email'];
    $message = (string) $validated['message'];
    $to = ichnm_feedback_recipient(['constant' => $recipient]);
    $mode = $mode === 'mail' ? 'mail' : 'journal';

    return [
        'ok' => true,
        'mode' => $mode,
        'action' => $mode,
        'recipient' => $to,
        'subject' => ICHNM_FEEDBACK_MAIL_SUBJECT,
        'body' => ichnm_feedback_mail_body($name, $email, $message),
        'name' => $name,
        'email' => $email,
        'message' => $message,
    ];
}

/**
 * Persist one feedback entry in the WordPress options journal (local honesty path).
 *
 * @param array{name:string,email:string,message:string} $payload
 */
function ichnm_feedback_write_journal(array $payload, bool $mail_sent = false): void
{
    $body = ichnm_feedback_mail_body($payload['name'], $payload['email'], $payload['message']);
    if (function_exists('error_log')) {
        error_log('[ichnm_feedback] ' . str_replace("\n", ' | ', $body));
    }
    if (function_exists('update_option')) {
        update_option('ichnm_last_feedback', [
            'time' => function_exists('current_time') ? current_time('mysql') : gmdate('Y-m-d H:i:s'),
            'name' => $payload['name'],
            'email' => $payload['email'],
            'message' => $payload['message'],
            'mail_sent' => $mail_sent,
            'mode' => $mail_sent ? 'mail' : 'journal',
        ], false);
    }
}

/**
 * Execute a prepared plan: journal XOR wp_mail.
 *
 * @param array{ok:true,mode:string,action:string,recipient:string,subject:string,body:string,name:string,email:string,message:string} $plan
 * @return array{ok:true,mode:string}|array{ok:false,error:string}
 */
function ichnm_feedback_execute(array $plan): array
{
    $payload = [
        'name' => $plan['name'],
        'email' => $plan['email'],
        'message' => $plan['message'],
    ];

    if ($plan['action'] === 'journal') {
        ichnm_feedback_write_journal($payload, false);
        return ['ok' => true, 'mode' => 'journal'];
    }

    $headers = ['Reply-To: ' . $plan['email']];
    $sent = function_exists('wp_mail')
        ? (bool) wp_mail($plan['recipient'], $plan['subject'], $plan['body'], $headers)
        : false;

    if (!$sent) {
        return ['ok' => false, 'error' => 'mail'];
    }

    ichnm_feedback_write_journal($payload, true);
    return ['ok' => true, 'mode' => 'mail'];
}
